### PR TITLE
make x-vault-token optional, allows using Vault Agent

### LINKED ISSUE #183: X-Vault-Token should be optional
Trying to add support for interacting with Vault agent: https://github.com/jenkinsci/hashicorp-vault-plugin/pull/36
Found out after much headache my issue was with vault-java-driver because it assumes vault-token is mandatory, which it is not in the case of the auto-auth use case for Vault agent. The null pointer results in 400 Bad Request.

image

Example of vault cli against the very same endpoint (ngrok used for inspection purpose :sweat:)

```
$ VAULT_ADDR=http://0c19b983.ngrok.io vault kv get kv-v1/admin
==== Data ====
Key     Value
---     -----
key1    123
key2    456
```

You can set up the vault agent to use auto-auth and handle the requesting the token. So the application does not need to be concerned about caching or reauthenticating against vault.

https://www.vaultproject.io/docs/agent/

### PR DESCRIPTION
**Pull request by casz**
fixes #183 

Added a vault agent test container to validate the fix.

### CODE REVIEW

**Comment by casz** [comment]
@steve-perkins @steve-perkins-bc anything I can help to speed this up? Do you need someone to help maintain the repo?

**Comment by casz** [comment]
I wonder if it would help to rewrite the `optionalHeader` to be the default implementation.
Currently, if `header` is given a null value the java library will throw 400 Bad request

File: src/main/java/com/bettercloud/vault/Vault.java [hunk 1]
```diff
@@ -249,7 +249,7 @@ public Seal seal() {
         try {
             final RestResponse restResponse = new Rest()//NOPMD
                     .url(vaultConfig.getAddress() + "/v1/sys/mounts")
-                    .header("X-Vault-Token", vaultConfig.getToken())
+                    .optionalHeader("X-Vault-Token", vaultConfig.getToken())
```
Comment (steve-perkins) [inline src/main/java/com/bettercloud/vault/Vault.java]: Huh.  It looks like another contributor added this `optionalHeader(...)` method back in February.  

If I thought about it at the time, I would have gone back and done all of this other cleanup to use it consistently (it's cleaner, and should have been like that to begin with).  Thanks for doing so.

**Comment by steve-perkins** [comment]
Thanks, @casz!

**Pull request merged**